spring-pulsar/src/main/java/org/springframework/pulsar/listener/DefaultPulsarMessageListenerContainer.java

```diff
@@ -439,6 +454,21 @@ public void run() {
 						handleAcks(messages);
 					}
 				}
+				if (isPaused()) {
```

Comment:
Here is the current flow in this PR:
```
1) get messages
2) if no messages then continue back to top of loop
3) there are messages handle batch or single
4) now, check if we are paused
```
When the condition for 2) matches it will go back to 1) and skip the pause check altogether. 

If we move the pause check to the top of the loop I think it solves this. Either that, or also do a "handlePause" in the  `messages is null` case.

Reply:
Very interesting. I didn't notice that flow where messages go null. I will update the PR.